drv/psc-seq-server/src/main.rs

```diff
@@ -3,48 +3,586 @@
+    Good,
+}
+
+struct Psu {
+    state: PsuState,
+}
+
+impl Psu {
+    /// Advances the PSU management state machine given the current time (`now`)
+    /// and the state of the `present` and `pwr_ok` inputs.
+    ///
+    /// This may be called at unpredictable intervals, and may be called more
+    /// than once for the same timestamp value. The implementation **must** use
+    /// `now` and the timer to control any time-sensitive operations.
+    fn step(
+        &mut self,
+        now: u64,
+        present: Present,
+        pwr_ok: Status,
+    ) -> Option<ActionRequired> {
+        match (self.state, present, pwr_ok) {
+            (PsuState::NotPresent, Present::No, _) => {
+                // ignore the power good line, it is meaningless.
+                None
+            }
+
+            // Regardless of our current state, if we observe the present line
+            // low, treat the PSU as having been disconnected.
+            //
+            // Other than detecting removal, the main side effect of this
+            // decision is that the "NewlyInserted" settle time starts after the
+            // contacts are _done_ scraping, not when they start.
+            (_, Present::No, _) => {
+                self.state = PsuState::NotPresent;
+                Some(ActionRequired::DisableMe {
+                    attempt_snapshot: false,
+                })
+            }
+
+            // In a not-present situation we have to ignore the OK line entirely
+            // and only watch for the presence line to indicate the PSU has
+            // appeared.
+            (PsuState::NotPresent, Present::Yes, _) => {
+                let settle_deadline = now.wrapping_add(INSERT_DEBOUNCE_MS);
+                self.state = PsuState::Present(PresentState::NewlyInserted {
+                    settle_deadline,
+                });
+                // No external action required until our timer elapses.
+                None
+            }
+
+            (
+                PsuState::Present(PresentState::NewlyInserted {
+                    settle_deadline,
+                }),
+                _,
+                _,
+            ) => {
+                if settle_deadline <= now {
+                    // The PSU is still present (since the Present::No case above
+                    // didn't fire) and our deadline has elapsed. Let's treat this
+                    // as valid!
+                    self.state = PsuState::Present(PresentState::On);
+                    Some(ActionRequired::EnableMe)
+                } else {
+                    // Remain in this state.
+                    None
+                }
+            }
+
+            // yay!
+            (PsuState::Present(PresentState::On), _, Status::Good) => None,
+
+            (PsuState::Present(PresentState::On), _, Status::NotGood) => {
+                // The PSU appears to have pulled the OK signal into the "not
+                // OK" state to indicate an internal fault!
```

Comment:
_very_ nitpicky, to the extent that i wasn't sure if i should even mention this: elsewhere, you've put the comment describing what case the states represent _above_ the match arm, but here it's under it:
```suggestion
            // The PSU appears to have pulled the OK signal into the "not  OK" 
            // state to indicate an internal fault!
            (PsuState::Present(PresentState::On), _, Status::NotGood) => {
```

Reply:
Interesting, these two classes of comments are very different things in my head, but not in a way I can easily explain or make actionable for future maintainers.

Basically, there's one class of comment explaining the _pattern,_ and one class explaining what to do after the pattern matches. e.g. the ones above the match arms are generally explaining why a field is removed or wildcard, the ones below talk strategy.

But I get that it just looks inconsistent, so I can make them all the same.

Reply:
Ah, interesting, I see what you mean! To me, the comments on the patterns seemed to describe, in a general sense, "what this match arm represents", so this felt more like that. But, it's up to you --- do whatever you think is best!